docs/content/_index.md

```diff
@@ -1,608 +1,1007 @@
+statically, or more importantly, the number of networks may not be known in
+advance.
+
+In Rego, the solution is to substitute the array index with a variable.
+
+```live:example/iter/1:query:merge_down
+some i; input.networks[i].public == true
+```
+```live:example/iter/1:output
+```
+
+Now the query asks for values of `i` that make the overall expression true. When
+you substitute variables in references, OPA automatically finds variable
+assignments that satisfy all of the expressions in the query. Just like
+intermediate variables, OPA returns the values of the variables.
+
+You can substitute as many variables as you want. For example, to find out if
+any servers expose the insecure `"http"` protocol you could write:
+
+```live:example/iter/double:query:merge_down
+some i, j; input.servers[i].protocols[j] == "http"
+```
+```live:example/iter/double:output
+```
+
+If variables appear multiple times the assignments satisfy all of the
+expressions. For example, to find the ids of ports connected to public networks,
+you could write:
+
+```live:example/iter/join:query:merge_down
+some i, j
+id := input.ports[i].id
+input.ports[i].network == input.networks[j].id
+input.networks[j].public
+```
+```live:example/iter/join:output
+```
+
+> If you split expressions across multiple lines you can omit the `;` operator.
+
+Providing good names for variables can be hard. If you only refer to the
+variable once, you can replace it with the special `_` (wildcard variable)
+operator. Conceptually, each instance of `_` is a unique variable.
+
+```live:example/iter/wildcard:query:merge_down
+input.servers[_].protocols[_] == "http"
+```
+```live:example/iter/wildcard:output
+```
+
+Just like references that refer to non-existent fields or expressions that fail
+to match, if OPA is unable to find any variable assignments that satisfy all of
+the expressions, the result is undefined.
+
+```live:example/iter/undefined:query:merge_down
+some i; input.servers[i].protocols[i] == "ssh"  # there is no assignment of i that satisfies the expression
+```
+```live:example/iter/undefined:output:expect_undefined
+```
+
+### Rules
+
+Rego lets you encapsulate and reuse logic with rules. Rules are just if-then
+logic statements. Rules can either be complete or incremental.
+
+#### Complete Rules
+
+Complete rules are if-then statements that assign a single value to a variable.
+For example:
+
+```live:example/complete/1:module:openable
+package example.rules
+
+any_public_networks = true {
+    net := input.networks[_]
+    net.public
 }
 ```
 
-As a result, any document, base or virtual, can be accessed hierarchically starting from the root data node – either as an identifier:
+Every rule consists of a _head_ and a _body_. In Rego we say the rule head is
+true _if_ the rule body is true for some set of variable assignments. In the
+example above `any_public_networks = true` is the head and `net :=
+input.networks[_]; net.public` is the body.
```

Comment:
Maybe add a note to try and change the input to have no public networks or something?